Add state class to UPnP sensors

aronbardsley opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds device class to the upnp component. This simply allows users to use the history graph statistics in lovelace. 
Refer to this article here: https://www.home-assistant.io/more-info/statistics/

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes N/A
- This PR is related to issue: N/A
- Link to documentation pull request: N/A

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [] The code has been formatted using Black (`black --fast homeassistant tests`)
- [] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

aronbardsley committed:
Update sensor.py

aronbardsley committed:
Update sensor.py

homeassistant commented:
Hi @aronbardsley, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#73782).

Once you do that we will be able to review and accept this pull request.

Thanks!

epenet commented on homeassistant/components/upnp/sensor.py:
```diff
@@ -149,6 +158,11 @@ class UpnpSensor(UpnpEntity, SensorEntity):
 class RawUpnpSensor(UpnpSensor):
     """Representation of a UPnP/IGD sensor."""
 
+    @property
+    def device_class(self):
+        """Return the state class for the sensor."""
+        return self.entity_description.state_class
+    
```
The property should not be needed. It should be automatically picked up from the `entity_description` by the base class.
https://github.com/home-assistant/core/blob/baa810aabbee430643229ee4b3e9f4c467d2878e/homeassistant/components/sensor/__init__.py#L301-L302

```suggestion
```

epenet commented on homeassistant/components/upnp/sensor.py:
```diff
@@ -175,6 +189,11 @@ def _has_overflowed(self, current_value) -> bool:
         """Check if value has overflowed."""
         return current_value < self._last_value
 
+    @property
+    def device_class(self):
+        """Return the state class for the sensor."""
+        return self.entity_description.state_class
+    
```
As above, the property should not be needed. It should be automatically picked up from the `entity_description` by the base class.

```suggestion
```

aronbardsley committed:
Update homeassistant/components/upnp/sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

aronbardsley committed:
Update homeassistant/components/upnp/sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

aronbardsley committed:
Merge branch 'dev' into upnp-patch-device-class

epenet reviewed (dismissed):
LGTM

frenck commented on homeassistant/components/upnp/sensor.py:
```diff
@@ -31,27 +31,31 @@
         icon="mdi:server-network",
         native_unit_of_measurement=DATA_BYTES,
         format="d",
+        state_class=SensorStateClass.MEASUREMENT,
```
Same as above

aronbardsley replied:
This has now been resolved.

frenck commented on homeassistant/components/upnp/sensor.py:
```diff
@@ -31,27 +31,31 @@
         icon="mdi:server-network",
         native_unit_of_measurement=DATA_PACKETS,
         format="d",
+        state_class=SensorStateClass.MEASUREMENT,
```
Same as above

aronbardsley replied:
This has now been resolved.

frenck commented on homeassistant/components/upnp/sensor.py:
```diff
@@ -65,6 +69,7 @@
         native_unit_of_measurement=TIME_SECONDS,
         entity_registry_enabled_default=False,
         format="d",
+        state_class=SensorStateClass.MEASUREMENT,
```
Is not a measurement. As a matter of fact, this sensor is actually against our guidelines, to begin with.

I suggest on removing this state class.

StevenLooman replied:
I plan on deprecating then removing the sensors, but I need to find a proper alternative first.

aronbardsley committed:
Merge branch 'dev' into upnp-patch-device-class

aronbardsley committed:
Resolve issues as per PR feedback

Remove state_class from sensors which cannot be measured (as they are over time)

BYTES_RECEIVED
BYTES_SENT
PACKETS_RECEIVED
PACKETS_SENT
PACKETS_SENT

aronbardsley commented on homeassistant/components/upnp/sensor.py:
```diff
@@ -31,27 +31,31 @@
         icon="mdi:server-network",
         native_unit_of_measurement=DATA_BYTES,
         format="d",
+        state_class=SensorStateClass.MEASUREMENT,
```
This has now been resolved.

frenck replied:
You've removed them? Why not set the right state_class instead?

aronbardsley commented on homeassistant/components/upnp/sensor.py:
```diff
@@ -31,27 +31,31 @@
         icon="mdi:server-network",
         native_unit_of_measurement=DATA_PACKETS,
         format="d",
+        state_class=SensorStateClass.MEASUREMENT,
```
This has now been resolved.

aronbardsley commented:
Thanks guys, my first PR on this software, I will do better next time.

StevenLooman commented:
> Thanks guys, my first PR on this software, I will do better next time.

I think you're doing great already! Just some minor changes are needed. You can still add commits/change things!

frenck commented:
@aronbardsley There is a merge conflict, could you take a look?

StevenLooman commented:
> I plan on deprecating then removing the [derived] sensors, but I need to find a proper alternative first.

I don't see that many options other than moving the sensors to the `async-upnp-client` library. Ultimately, this doesn't change that much other than moving the same logic around.

Ideally, I'd use a template sensor for this. But then it would require to get the previous state value and previous timestamp of that state (value) to calculate the change per second value.

What do you propose @frenck?

StevenLooman commented:
> > I plan on deprecating then removing the [derived] sensors, but I need to find a proper alternative first.
> 
> I don't see that many options other than moving the sensors to the `async-upnp-client` library. Ultimately, this doesn't change that much other than moving the same logic around.
> 
> Ideally, I'd use a template sensor for this. But then it would require to get the previous state value and previous timestamp of that state (value) to calculate the change per second value.
> 
> What do you propose @frenck?

I caved in and moved the calculation of the derived measurements to `async-upnp-client`.

StevenLooman commented:
I'm closing this PR. #79955 supersedes this. Please, do give @aronbardsley credits when the other PR is merged!
